### PR TITLE
api: fix Derivative dict substitution with self as key

### LINKED ISSUE #1995: Derivatives get duplicated when a derivative is substituted for expression containing derivatives
If an unevaluated derivative is substituted for some expression containing derivatives, any derivatives in the substituted expression are duplicated unless they are the original derivative. As the `Derivative` object gets duplicated, subsequent algebraic simplifications are missed during the lowering process as they are considered separate symbols and thus not canceled.

If this expression is then used in building a kernel, it will take much longer than usual to build as a result of the increased number of derivative objects.

MFE:
```
from devito import Grid, Function, Derivative

grid = Grid(shape=(11,), extent=(10.,))
f = Function(name='f', grid=grid, space_order=4)

# To make sure that derivative objects are the same
fdx = f.dx
fdx2 = f.dx2

expr = fdx + fdx2

assert expr.find(Derivative) == expr.subs({fdx: fdx + fdx2, fdx2: fdx2}).find(Derivative)
```

### PR DESCRIPTION
**Pull request by mloubout**
Fixes #1995

### CODE REVIEW

**Comment by mloubout** [comment]
@EdCaunt ok so this is not as minor as anticipated. Mainly the cause of the issue you raised is that, unless you do `subs(self, new)` then the substitution is delayed and the dictionary stored as a property. it means if you do `f.dx.subs({'a': 2})` then it will keep that dict and it will be part of the `==` comparison.


So in your case you end up with two different `fdx`. The original one and a new one that is "fdx but remember {fdx2: ....}" which even though in this case won't do anything leads to it being a different object. There isn't really an easy way out of it because some of the keys may not exist until after evaluation so there is no way to know if that dict needs to be saved to not.

So to summarize,  your issue is just a side effect of the lazy evaluation and there is no straightforward way to get out of it.

PS: renaming as WIP and draft cuz may not be doable nicely.

File: devito/finite_differences/derivative.py [hunk 1]
```diff
@@ -213,6 +213,14 @@ def subs(self, *args, **kwargs):
         # because a new derivative could be created i.e `f.dx.subs(f.dx, y)`
         if len(args) == 2 and args[0] == self:
             return args[1]
+        # Check if we are calling subs(self, {old: new}, **hint) in which case
+        # return the standard substitution.
+        elif len(args) == 1 and self in args[0]:
+            new = args[0].pop(self)
```
Comment (FabioLuporini) [inline devito/finite_differences/derivative.py:219]: just use `self`?

File: devito/finite_differences/derivative.py [hunk 2]
```diff
@@ -213,6 +213,14 @@ def subs(self, *args, **kwargs):
         # because a new derivative could be created i.e `f.dx.subs(f.dx, y)`
         if len(args) == 2 and args[0] == self:
```
Comment (FabioLuporini) [inline devito/finite_differences/derivative.py:214]: is self

File: devito/finite_differences/derivative.py (hunk 1 above)
Comment (mloubout) [reply devito/finite_differences/derivative.py:219]: No this is `args[0][self]` not `self`. I.e this extract what to replace `self` with.
